File: Intersection of Two Linked Lists/ITLL.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def getIntersectionNode(self, headA, headB):
        """
        :type head1, head1: ListNode
        :rtype: ListNode
        """
        if headA==None or headB==None:return None
        hA = headA
        hB = headB
        lenA = 0
        lenB = 0
        while hA!=None:
            lenA+=1
            hA = hA.next
        while hB!=None:
            lenB+=1
            hB = hB.next
        hB = headA
        hA = headB
        if lenB>lenA:
            for i in range(0,lenB-lenA):
                hA = hA.next
        if lenA>lenB:
            for i in range(0,lenA-lenB):
                hB = hB.next
        while hA!=None and hB!=None:
            if hA==hB: return hA
            hA=hA.next
            hB=hB.next
        return None
            

# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def getIntersectionNode(self, headA, headB):
        """
        :type head1, head1: ListNode
        :rtype: ListNode
        """
        if headA==None or headB==None:return None
        hA = headA
        hB = headB
        while hA!=hB:
            hA = hA.next if hA!=None else headB
            hB = hB.next if hB!=None else headA
        return hA

File: Intersection of Two Linked Lists/ITLL.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode getIntersectionNode(ListNode headA, ListNode headB) {
        if(headA ==null || headB==null) return null;
        ListNode hA = headA;
        ListNode hB = headB;
        int lenA = 0;
        int lenB = 0;
        while(hA.next!=null){
            lenA++;
            hA = hA.next;
        }
        while(hB.next!=null){
            lenB++;
            hB = hB.next;
        }
        hA = headA;
        hB = headB;
        if(lenB>lenA){
            int i = 0;
            while (i<lenB-lenA){
                hB = hB.next;
                i++;
            }
        }
        if(lenB<lenA){
            int i = 0;
            while (i<lenA-lenB){
                hA = hA.next;
                i++;
            }
        }
        while(hA!=null && hB!=null){
            if(hA == hB) return hA;
            hA = hA.next;
            hB = hB.next;
        }
        return null;
    }
}

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode getIntersectionNode(ListNode headA, ListNode headB) {
        if(headA==null || headB==null) return null;
        ListNode h1 = headA; 
        ListNode h2 = headB;
        while(h1!=h2){
            h1 = h1 == null?headB:h1.next;
            h2 = h2 == null?headA:h2.next;
        }
        return h1;
    }
}